### PR TITLE
Redesigned Python API

### PR DESCRIPTION
**Pull request by sukritkalra**

### CODE REVIEW

File: python/erdos/streams.py
Comment (ICGog) [inline python/erdos/streams.py:216]: Typo

**Review by ICGog — COMMENTED** [review]
Comments from a partial review.

File: python/erdos/timestamp.py
Comment (sukritkalra) [reply python/erdos/timestamp.py]: Fixed.

File: python/erdos/operator.py
Comment (pschafhalter) [inline python/erdos/operator.py:9]: Can the classes in this module inherit from [abstract base class](https://docs.python.org/3/library/abc.html)?

File: python/erdos/operator.py
Comment (pschafhalter) [inline python/erdos/operator.py:83]: API question: should we force the user to implement all of these methods when writing new operators (e.g. by raising `NotImplementedError` or marking them as abstract methods?

An advantage is that the system will catch bugs, e.g. from mispelling the method name when overriding. A disadvantage is that a new operator must implement all of these methods.

File: python/erdos/timestamp.py
Comment (pschafhalter) [inline python/erdos/timestamp.py:53]: Might be easier to maintain if we always instantiate a `PyTimestamp`, and store `self._py_timestamp`.

Then we can re-use the Rust implementation for `__lt__`, `__leq__` by writing it in Rust in the `PyTimestamp` class. Properties like `is_top` would return `self._py_timestamp.is_top`, etc.

Comment (pschafhalter) [reply python/erdos/timestamp.py:53]: We might also want to consider moving the `Timestamp` class to pure Rust and getting rid of this Python wrapper. I believe pyo3 allows us to add docstrings from Rust now too.

Thoughts?

File: src/python/py_timestamp.rs
Comment (pschafhalter) [inline src/python/py_timestamp.rs:68]: It's [better practice](https://doc.rust-lang.org/std/convert/trait.From.html) to `impl From<PyTimestamp> for Timestamp`

**Review by pschafhalter — COMMENTED** [review]
Why do we need separate Python executors? Can't we implement `PySource`, `PySink`, `PyOneInOnOut`, etc. operators? These would take a reference to the Python operator as an argument, and be responsible for calling the correct methods on the Python operator object.

In addition, this would also let us use the Rust `connect_*` functions, rather than duplicating much of the logic in `python/mod.rs`.

I'm in general pretty concerned about the amount of logic duplication going on between the Python and Rust APIs. A lot of duplication makes it difficult to maintain the codebase and introduce new features.

File: src/python/py_timestamp.rs
Comment (sukritkalra) [reply src/python/py_timestamp.rs:68]: Thanks! I'll change this to a `From` in this file instead of having it in `src/dataflow/time.rs` though.

File: python/erdos/__init__.py
Comment (sukritkalra) [reply python/erdos/__init__.py:423]: Yes, added them in the latest changes.

File: python/erdos/timestamp.py
Comment (sukritkalra) [reply python/erdos/timestamp.py]: I wrote a minor PyTimestamp wrapper that patches almost all methods from Python to Rust so we utilize our implementations in Rust now for comparisons.

File: python/erdos/operator.py
Comment (sukritkalra) [reply python/erdos/operator.py:88]: I added a logger warning if the user leaves all of the methods of an operator empty.

File: Cargo.toml
Comment (sukritkalra) [inline Cargo.toml:30]: I need to bump this before merging.

**Review by pschafhalter — APPROVED** [review]
LGTM, once the merge conflict is fixed and PyO3 is bumped to 0.14.3.

**Pull request merged**
